### DISCUSSION TITLE
`enquire_link` interval - manually change - how to?

### DISCUSSION THREAD

**Discussion by Bosnianflajter2** [python-smpplib/python-smpplib#199] [Q&A] [answered]
Hello,

How can I manually change the send interval of `enquire_link`?

```
DEBUG:smpp.Client.1649662428208:Sending enquire_link PDU
DEBUG:smpp.Client.1649662428208:>>b'00000010000000150000000000000001' (16 bytes)
DEBUG:smpp.Client.1649662428208:Waiting for PDU...
DEBUG:smpp.Client.1649662428208:<<b'00000010800000150000000000000001' (16 bytes)
DEBUG:smpp.Client.1649662428208:Read enquire_link_resp PDU
DEBUG:smpp.Client.1649662428208:Waiting for PDU
```

Currently an enquire_link is sent every 2-3 seconds.
And the third-party SMPP server responded correctly.
But I need to set the `enquire_link` interval on the client to send every 30 or 60 seconds.
How can I do that?

thx
best

**Comment by eigenein** [comment, marked as answer]
Hi,

The client should only send `enquire_link` when the underlying socket times out.

Do you have any logging entries right before the `Sending enquire_link PDU`?

**Reply by Bosnianflajter2** [reply]
Hello eigenein,

Thx for the response.

Here the stdout right aftert starting of the smpp client:

INFO:smpp.Client.2399640159280:Connecting to 213.190.x.x:2875...
DEBUG:smpp.Client.2399640159280:Receiver mode
DEBUG:smpp.Client.2399640159280:Sending bind_transceiver PDU
DEBUG:smpp.Client.2399640159280:>>b'000000280000000900000000000000016d6d616c6261736963004d73472e53213137000034000000' (40 bytes)
DEBUG:smpp.Client.2399640159280:Waiting for PDU...
DEBUG:smpp.Client.2399640159280:<<b'00000019800000090000000000000001706f736569646f6e00' (25 bytes)
DEBUG:smpp.Client.2399640159280:Read bind_transceiver_resp PDU
DEBUG:smpp.Client.2399640159280:Waiting for PDU...
DEBUG:smpp.Client.2399640159280:Socket timeout, listening again
DEBUG:smpp.Client.2399640159280:Sending enquire_link PDU
DEBUG:smpp.Client.2399640159280:>>b'00000010000000150000000000000001' (16 bytes)
DEBUG:smpp.Client.2399640159280:Waiting for PDU...
DEBUG:smpp.Client.2399640159280:<<b'00000010800000150000000000000001' (16 bytes)
DEBUG:smpp.Client.2399640159280:Read enquire_link_resp PDU
DEBUG:smpp.Client.2399640159280:Waiting for PDU...
DEBUG:smpp.Client.2399640159280:Socket timeout, listening again
....
....

And this is in the SMPP Server log from third party:

[2022-04-25 22:36:30.09576] [root SMPP-Server-v3.1 BosnianSRV 108610] [INFO]  connection from: 10.132.x.x
[2022-04-25 22:36:30.09775] [root SMPP-Server-v3.1 BosnianSRV 108610] [INFO]  (testuser) processing bind
[2022-04-25 22:36:30.09836] [root SMPP-Server-v3.1 BosnianSRV 108610] [INFO]  (testuser) remote version = 52
[2022-04-25 22:36:30.10157] [root SMPP-Server-v3.1 BosnianSRV 108610] [INFO]  registering connection [transceiver]= 1
[2022-04-25 22:36:30.10205] [root SMPP-Server-v3.1 BosnianSRV 108610] [INFO]  (testuser) bind: transceiver successful [10.132.x.x,52936]  (52)
....
....


You're right, the client output says "...Socket timeout, listening again...".
But... why are the timeouts coming?
Does this mean that every x seconds the 3rd party SMPP server will freeze/drop the existing smpp connection and therefore the smpp client will try to reconnect using enq_link?
How can I avoid this?
Because I have a response from 3rd party SMPP server saying I am sending enq_link every 4-5 seconds and I need to increase the intervals to not "overwhelm" the SMPP server.


Maybe you have some input for me how to get it increased to at least every 30 sec sent.

thx again.
best

**Reply by eigenein** [reply]
> But... why are the timeouts coming?
> Does this mean that every x seconds the 3rd party SMPP server will freeze/drop the existing smpp connection and therefore the smpp client will try to reconnect using enq_link?

It's difficult to explain why the timeouts occur having only this much information at hand, but you're basically correct: there's something wrong with the underlying socket. You may try increasing the `timeout` parameter of the `Client` – by default it's set to 5 seconds. Perhaps this will help you overcome eventual network freezes/errors and there will be no need to reconnect at all. I'd go with `10`, `20`, `30` seconds.

As the last resort you may try setting `auto_send_enquire_link=False` for the `listen()` call to disable the `enquire_link` calls – but I wouldn't recommend that unless you understand what you're doing.

**Reply by Bosnianflajter2** [reply]
thx eigenein,

I will try that tmrw during the day.

best

**Reply by Bosnianflajter2** [reply]
I was not able to find  out where to set the enq_link interval.
Investigated the code (all *py files in the smpp lib) and nowhere is a value set.
I was expecting some integer / interval value what I could change / set to 30 (instead of 5 seconds).
Could somebody post me here the line number (and *.py file from the smpp lib) where this can be changed?

thx

**Reply by eigenein** [reply]
Well, because there's no such setting. As I said before:

> The client should only send enquire_link when the underlying socket times out.

and suggested to try and change the `Client`-s `timeout` parameter.

There's nothing else I can answer on the same question asked twice ;)

**Reply by Bosnianflajter2** [reply]
thx eigenein,
Sorry to botter you again.
Maybe I didn't catch that corerct. xD

Ok, found that and changed to 30
" timeout=30," in Client.py

Obviously the SMPP Server is sending the enq_links now as my client do answer only (enquire_link_resp):
....
DEBUG:smpp.Client.1906893586240:Waiting for PDU...
DEBUG:smpp.Client.1906893586240:<<b'0000001000000015000000000000000b' (16 bytes)
DEBUG:smpp.Client.1906893586240:Read enquire_link PDU
DEBUG:smpp.Client.1906893586240:Sending enquire_link_resp PDU
DEBUG:smpp.Client.1906893586240:>>b'0000001080000015000000000000000b' (16 bytes)
DEBUG:smpp.Client.1906893586240:Waiting for PDU...
....

Which is also a workaround and in my case OK now. xD

@eigenein: thx for your help.
I think we can close that case as resolved.

Wish you good night - at least here in Vienna is 10:42pm now.

bye

**Reply by eigenein** [reply]
Right, the server is supposed to send them from time to time. And now you don't read timeouts, right? Then, it's more a solution than a workaround – you just let the server be less responsive while still maintaining a stable link, which should be fine.
